docs/fobnail-api.md

```diff
@@ -0,0 +1,494 @@
+                ...
+            ]
+        },
+        // Another PCR bank (major 5, map)
+        {
+            "algo_id": 0xb,
+            "pcrs": 0x000000ff,
+            // major 4 (array)
+            "pcr": [
+                // major 2 (byte string)
+                [ 0x00, 0x00, 0x00, 0x00, ... ],
+                ...
+            ]
+        }
+    ],
+}
+```
+
+Client **must** sign RIM with AIK (see the [Data signing](#data-signing) section
+above).
+
+Fobnail Token verifies RIM signature and format. If signature verification
+fails, **4.03** is returned. If RIM has an invalid format, **4.00** is returned.
+On success **2.01** or **2.04** is returned. On the first successful call,
+**2.01** is returned (`Location-Path` must not be present); on subsequent calls,
+**2.04** is returned to signify that the previous RIM has been replaced with the
+new RIM.
+
+#### /admin/provision/{id}
+
+POST request to this endpoint results in writing data attached to the
+Provisioning Context into persistent storage. Currently, this request does not
+contain any data, and payload **must** be empty or **4.00** response will be
+returned.
+
+Fobnail Token verifies whether all required data has been provided (metadata and
+RIM is mandatory). On error, **4.03** response is returned. If verification is
+successful, Fobnail Token writes all required data into persistent storage,
+completing the provisioning process. If an error occurs while writing into
+persistent storage, **5.00** is returned (the payload may contain additional
+context). On success, **2.04** response is returned to signify provisioning is
+complete. At this point, the Provisioning Context becomes invalid, and all
+subsequent requests to the Provisioning Context will return **4.04**.
+
+### Attestation
+
+| Endpoint Name             | Method | Arguments            |
+| ------------------------- | ------ | -------------------- |
+| /attest                   | POST   | Platform metadata    |
+| /attest/{id}              | POST   | Output of TPM quote  |
+
+#### /attest
+
+The client sends platform metadata in the same format as when calling
+`/admin/provision/{id}/meta`. Metadata **must** be signed by the same AIK that
+was used during platform provisioning. If verification is successful **2.01** is
+returned (otherwise **4.04**) and `Location-Path` contains Attestation Context
+ID. The response's payload contains PCR selection and nonce:
+
+```json
+{
+    // Array of PCR banks (major 4)
+    "banks": [
+        // First PCR bank
+        {
+            // ID of algorithm used by PCRs (major 0)
+            "algo_id": 0x0004,
+            // Bitmap of present PCRs (major 0, unsigned integer)
+            "pcrs": 0xffffffff,
+        },
+        // Another PCR bank (major 5, map)
+        {
+            "algo_id": 0x000b,
+            "pcrs": 0x000000ff,
+        }
+    ],
+    // Nonce to be passed to TPM_Quote() (major 2)
+    "nonce": [0x00, 0x01, 0x02, 0x03, ...]
```

Comment:
This partially overlaps with new `/nonce` endpoint. I'm OK with keeping it in the structure to minimize number of requests, but in that case we need to strictly define whether `/nonce` is allowed between `/attest` and `/attest/{id}`.

I'd say it should be forbidden, and result in invalidation of attestation context ID, in preparation for another `/attest` message. Otherwise compromised client can keep requesting new `/nonce` until it receives a nonce that was used in previous quote and replay that message.

Reply:
Done